ships/0004-build-env-vars.md

```diff
@@ -0,0 +1,228 @@
+
+### Non-Goals
+
+* Allow developers to alter or override other aspects of the build step, such as commands.
+* Set environment variables in the output image of the build.
+* Validate environment variables against those in the base image of a build step.
+* Provide direct support for Dockerfile [build arguments](https://docs.docker.com/engine/reference/commandline/build/#set-build-time-variables---build-arg).
+
+## Proposal
+
+### User Stories
+
+#### Story: Alter build behavior via environment variables
+
+As a nodejs developer using buildpacks
+I want to add `NPM_*` environment variables to my Build and BuildRuns
+So that I can tune the behavior of npm in my build
+
+#### Story: Prevent insecure behavior in a build strategy
+
+As a build strategy author
+I want to keep fixed values for some environment variables in my build strategy
+So that I don't allow end users to enable insecure or potentially harmful features.
+
+### Implementation Notes
+
+#### Build Strategy Behavior
+
+Build strategies which declare an environment variable in a build strategy step will by default be fixed.
+A developer will not be able to change this value directly.
+
+If a build strategy author wants to allow an environment variable to be tuned, they have two options:
+
+1. Omit the environment variable - this works if an empty/zero default value is acceptable in the build step.
+2. Declare the envionment variable, but use a build strategy parameter to set the value.
+   This is useful if the build strategy should use a nonempty/nonzero default value in the build step.
+
+For example, a BuildKit build strategy author could use the following to ensure that TLS is always verified when pushing the resulting image:
+
+```yaml
+spec:
+  steps:
+  - name: build
+    ...
+    env:
+    - name: DOCKER_TLS_VERIFY
+      value: "true"
+```
+
+Or alternatively, the build strategy author can make this configurable with a build parameter:
+
+```yaml
+spec:
+  params:
+  - name: TLS_VERIFY
+    description: "Verify TLS when pushing the resulting container image."
+    default: "true"
+  steps:
+  - name: build
+    ...
+    env:
+    - name: DOCKER_TLS_VERIFY
+      value: $(params.TLS_VERIFY)
+```
+
+#### Build/BuildRun Environment Variables
+
+The `Build` and `BuildRun` APIs will be extended with a `spec.env` field.
+This is where a developer declares that they are appending environment variables to a build step.
+For this initial implementation, `env` will have an object map of step name to environment variables to be added.
+
+Referring to the nodejs buildpacks story above, a developer who wants their build to install `dev` dependencies would do the following in their `Build` or `BuildRun`:
+
+```yaml
+spec:
+  sources:
+  - name: git
+    type: Git
+    git:
+      uri: https://github.com/sclorg/nodejs-ex.git
+  strategy:
+    kind: BuildStrategy
+    name: buildpacks-v3
+  env:
+    build:
```

Comment:
I fully understand the rational here, which is to avoid complexity around validating which env vars apply to which strategy step, and which not. But I´m not sure that complexity is even needed. 

**Reasons on why that complexity is not needed:**

- An env var defined at the `Build` or `BuildRun` doesnt necessarily mean that it will reach the `corev1.Container.Env` field of a strategy step.
- A user can define N env vars, but they are irrelevant as long as the step container image does not use it.
- Strategy authors know in advance which env vars can modify the behaviour of the image execution. If they know on X env var that users should not be changing, then they need to explicitely define that env var `name:value` under the strategy step ( _to block users from setting them_ ). Otherwise, they allow a malicious user to take advantage of that.
- Strategy authors have the responsability to document which are the env vars that the strategy supports and why, independently of the step. This is a responsibility that we havent highlight before in our docs, as we didnt have the need, but today with the parameters and env vars in place, we should do this. There are multiple ways on how to achieve this, for example:
   - in Shipwright/Build we should have a docs on what can we parameterize for our BuildStrategies
   - in workflows like CLI, we should have a command that provide this information

**Reasons on why I would like to avoid having specific strategy details information in the Build:**

- We do not follow this approach for the paremeterization [implementation](https://github.com/shipwright-io/build/pull/781), doing this approach for envs will confuse users. As they will be able to define params per strategy, but then env vars per strategy step.
- As explained above in `Reasons on why that complexity is not needed`, I believe is fine to allow users to define env vars that might apply to N steps on the strategies. Strategy authors should be able to signalize where to use them and if they can be used or not.
- Adding the notion of a step name in the `Build` / `BuildRun`breaks the simplicity of the relationship between these objects and the Strategies. Let me elaborate more on this:
   - Today, all what users need to know is which strategy to choose ( name and type ). In order to choose wisely, they need to read a little bit on the docs, or whatever means admins have to provide this information for their users.
   - Tomorrow, users need to know which strategy to use ( name and type ), but also which parameters are supported in that strategy and how this modifies the strategy behaviour ( _more docs reading_ ). Note: They do not need to know on a specific step, they just need to know that if param Z is defined, then the image will go to a "local container registry".
   - For env vars, we should avoid forcing users to know on which exact `step` the variable will be use, they should just know that there is this A,B env var that I could set, and this "behaviour" will take place.

Reply:
> For env vars, we should avoid forcing users to know on which exact step the variable will be use, they should just know that there is this A,B env var that I could set, and this "behaviour" will take place.

I share this concern. I may be too conservative in my hypothetical edge case where an env var declaration in one step can break a subsequent step. Furthermore, by applying the env var to every step we make life simpler for proxied clusters, where `*_PROXY` env vars need to be injected into every step of the build.